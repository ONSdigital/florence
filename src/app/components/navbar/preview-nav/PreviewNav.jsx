import React from "react";
import { push } from "react-router-redux";
import PropTypes from "prop-types";
import { updateSelectedPreviewPage } from "../../../config/actions";
import Select from "../../Select";

const PreviewNav = props => {
    const mapPagesToSelect = pages => {
        if (pages) {
            try {
                console.log(pages);
                return pages.map(page => {
                    if (page.type === "visualisation" && page.files) {
                        return {
                            id: page.uri,
                            name: createPageTitle(page),
                            isGroup: true,
                            groupOptions: page.files.map(file => {
                                return { id: `${page.contentPath}/${file}`, name: file };
                            })
                        };
                    }
                    return {
                        id: page.uri,
                        name: createPageTitle(page)
                    };
                });
            } catch (err) {
                console.error("Error mapping pages to select", err);
            }
        }
        return;
    };

    const createPageTitle = page => {
        if (page.description.title && page.description.edition) {
            return `${page.description.title}: ${page.description.edition}`;
        }
        if (!page.description.title && page.description.edition) {
            return `[no title available]: ${page.description.edition}`;
        }
        if (page.description.title && !page.description.edition) {
            return page.description.title;
        }
        return "";
    };

    const handleSelectChange = event => {
        const selection = event.target.value;
        if (selection === "default-option") {
            return;
        }
        const uri = selection;
        props.dispatch(updateSelectedPreviewPage(uri));
        push(`${props.rootPath}/collections/${props.workingOn.id}/preview?url=${uri}`);
    };

    return (
        <div className="global-nav__preview-select">
            <Select
                id="preview-select"
                contents={mapPagesToSelect(props.preview.pages) || []}
                onChange={handleSelectChange}
                defaultOption={props.preview.pages ? "Select an option" : "Loading pages..."}
                selectedOption={props.preview.selectedPage || ""}
            />
        </div>
    );
};

PreviewNav.propTypes = {
    dispatch: PropTypes.func.isRequired,
    preview: PropTypes.object.isRequired,
    rootPath: PropTypes.string.isRequired,
    workingOn: PropTypes.object.isRequired
};

export default PreviewNav;
