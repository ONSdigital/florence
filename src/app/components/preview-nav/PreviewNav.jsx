import React, { useEffect, useState } from "react";
import { push } from "react-router-redux";
import PropTypes from "prop-types";
import Select from "../Select";

const PreviewNav = ({ workingOn, preview, rootPath, updateSelected }) => {
    if (!workingOn) return null;
    const [uri, setUri] = useState("/");

    useEffect(() => {
        updateSelected(uri);
        push(`${rootPath}/collections/${workingOn.id}/preview?url=${uri}`);
    }, [uri]);

    const mapPagesToSelect = pages => {
        if (pages) {
            try {
                return pages.map(page => {
                    if (page.type === "visualisation" && page.files) {
                        return {
                            id: page.uri,
                            name: createPageTitle(page),
                            isGroup: true,
                            groupOptions: page.files.map(file => {
                                return { id: `${page.contentPath}/${file}`, name: file };
                            }),
                        };
                    }
                    if (page.type === "interactive" && page.files) {
                        return {
                            id: page.uri,
                            name: createPageTitle(page),
                            isGroup: true,
                            groupOptions: page.files.map(file => {
                                return { id: file.uri, name: file.uri };
                            }),
                        };
                    }
                    return {
                        id: page.uri,
                        name: createPageTitle(page),
                    };
                });
            } catch (err) {
                console.error("Error mapping pages to select", err);
            }
        }
        return;
    };

    const createPageTitle = page => {
        if (page.description.title && page.description.edition) {
            return `${page.description.title}: ${page.description.edition}`;
        }
        if (!page.description.title && page.description.edition) {
            return `[no title available]: ${page.description.edition}`;
        }
        if (page.description.title && !page.description.edition) {
            return page.description.title;
        }
        return "";
    };

    return (
        <div className="global-nav__preview-select">
            <Select
                id="preview-select"
                contents={mapPagesToSelect(preview.pages) || []}
                onChange={e => setUri(e.target.value)}
                defaultOption={preview.pages ? "" : "Loading pages..."}
                selectedOption={uri}
            />
        </div>
    );
};

PreviewNav.propTypes = {
    preview: PropTypes.object.isRequired,
    rootPath: PropTypes.string.isRequired,
    workingOn: PropTypes.object.isRequired,
};

export default PreviewNav;
