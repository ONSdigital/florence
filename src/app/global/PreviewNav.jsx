import React, { Component } from 'react';
import { connect } from 'react-redux';
import { push } from 'react-router-redux';
import PropTypes from 'prop-types';

import { updateSelectedPreviewPage } from '../config/actions'

import Select from '../components/Select'

const propTypes = {
    dispatch: PropTypes.func.isRequired,
    preview: PropTypes.object.isRequired
}

class PreviewNav extends Component {
    constructor(props) {
        super(props);

        this.handleSelectChange = this.handleSelectChange.bind(this);
    }

    mapPagesToSelect(pages) {
        if (pages) {
            try {
                return pages.map(page => {
                    return {
                        id: page.uri,
                        name: page.description.title
                    }
                })
            } catch (err) {
                console.error("Error mapping pages to select", err)
            }
        }
        return;
    }

    handleSelectChange(event) {
        const uri = event.target.value;
        this.props.dispatch(updateSelectedPreviewPage(uri));
        this.props.dispatch(push(`${window.location.pathname}?url=${uri}`));
    }

    render() {
        return (
            <span>
                <Select id="preview-select" 
                    contents={this.mapPagesToSelect(this.props.preview.pages) || []} 
                    onChange={this.handleSelectChange}
                    defaultOption={this.props.preview.pages ? "Select an option" : "Loading pages..."}
                    selectedOption={this.props.preview.selectedPage || "" }
                />
            </span>
        );
    }
}

PreviewNav.propTypes = propTypes;

export function mapStateToProps(state) {
    return {
        preview: state.state.preview,
    }
}

export default connect(mapStateToProps)(PreviewNav);