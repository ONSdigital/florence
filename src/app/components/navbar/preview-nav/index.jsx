import { connect } from "react-redux";
import PreviewNav from "./PreviewNav";
import { updateSelectedPreviewPage } from "../../../config/actions";

const mapStateToProps = state => {
    return {
        preview: state.state.preview,
        rootPath: state.state.rootPath,
        workingOn: state.state.global ? state.state.global.workingOn : {}
    };
};

const mapDispatchToProps = dispatch => {
    return {
        updateSelectedPreviewPage: uri => dispatch(updateSelectedPreviewPage(uri))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(PreviewNav);
