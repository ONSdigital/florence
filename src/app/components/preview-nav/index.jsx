import { connect } from "react-redux";
import PreviewNav from "./PreviewNav";

function mapStateToProps(state) {
    const user = state.state.user;
    const rootPath = state.state.rootPath;
    const workingOn = state.state.global ? state.state.global.workingOn : null;

    return {
        user,
        rootPath,
        workingOn,
        config: state.state.config,
    };
}

export default connect(mapStateToProps)(PreviewNav);
