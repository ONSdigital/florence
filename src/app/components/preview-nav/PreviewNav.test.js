import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import { mapStateToProps } from "../preview-nav";
import PreviewNav from "../preview-nav/PreviewNav";

const pages = [
    {
        uri: "/test-uri",
        type: "bulletin",
        description: {
            title: "Test title",
            edition: "Edition",
            language: "English",
        },
    },
    {
        uri: "/test-uri2",
        type: "bulletin",
        description: {
            title: "Test title 2",
            edition: "",
        },
    },
    {
        uri: "/test-uri3",
        type: "bulletin",
        description: {
            title: "",
            edition: "Edition 3",
        },
    },
];

const defaultProps = {
    preview: {
        selectedPage: null,
        collectionID: "test-collection-123",
        name: "My collection",
    },
    workingOn: {
        id: "test-collection",
    },
    rootPath: "/florence",
    updateSelected: () => {},
};

const withPagesProps = {
    ...defaultProps,
    preview: {
        ...defaultProps.preview,
        pages: pages,
    },
};

describe("PreviewNav", () => {
    it("should match the snapshot", async () => {
        const tree = renderer.create(<PreviewNav {...defaultProps} />);
        expect(tree).toMatchSnapshot();
    });

    it("should map state to props", () => {
        const state = {
            state: {
                preview: {},
                rootPath: "/foo",
                global: {
                    workingOn: { id: "baz" },
                },
            },
        };

        const props = {
            preview: {},
            rootPath: "/foo",
            workingOn: { id: "baz" },
        };

        const componentState = mapStateToProps(state, props);
        expect(componentState).toEqual(props);
    });

    it("renders select with default option selected", () => {
        const wrapper = mount(<PreviewNav {...defaultProps} />);
        expect(wrapper.find(".select").text()).toEqual("Loading pages...");
    });

    it("generates page titles in Select options", () => {
        const wrapper = mount(<PreviewNav {...withPagesProps} />);
        const options = wrapper.find("option");

        expect(options.at(0).text()).toEqual("Select an option");
        expect(options.at(1).text()).toEqual("Test title: Edition");
        expect(options.at(2).text()).toEqual("Test title 2");
        expect(options.at(3).text()).toEqual("[no title available]: Edition 3");
    });

    it("allows to select page", () => {
        const wrapper = mount(<PreviewNav {...withPagesProps} />);
        expect(wrapper.find("select").props().value).toBe("/");
        wrapper.find("select").simulate("change", { target: { value: "/test-uri" } });
        expect(wrapper.find("select").props().value).toBe("/test-uri");
    });
});

// TODO: when added react-router-dom to use mock location
// it("routes to selected page on page selection", () => {
//   component.handleSelectChange({ target: { value: "/test-uri" } });
//   expect(dispatchedActions[1].type).toBe("@@router/CALL_HISTORY_METHOD");
//   expect(dispatchedActions[1].payload.method).toBe("push");
//   expect(dispatchedActions[1].payload.args[0]).toBe("/florence/collections/test-collection/preview?url=/test-uri");
// });
// });
