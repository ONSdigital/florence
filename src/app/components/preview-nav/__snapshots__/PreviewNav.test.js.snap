// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewNav should match the snapshot 1`] = `
<div
  className="global-nav__preview-select"
>
  <div
    className="form__input"
  >
    <label
      className="form__label"
      htmlFor="preview-select"
    />
    <div
      className="select-wrap "
    >
      <select
        className="select"
        disabled={false}
        id="preview-select"
        name="preview-select"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        value="/"
      >
        <option
          value="Loading pages..."
        >
          Loading pages...
        </option>
      </select>
    </div>
  </div>
</div>
`;
